API review:  Add Workers management APIs. (#4556) (#4682)
MicrosoftEdge/WebView2Feedback: specs/Workers.md

Workers support in WebView2
===

# Background
Currently, although workers can be used in web content in WebView2, WebView2
lacks comprehensive WebView2 APIs for developers to fully utilize workers,
leading to increased load on the main thread, offloading to native, or
forking your web app for WebView2 scenarios. To address this, the WebView2 team
is introducing support for worker scenarios. This enhancement allows WebView2
users to interact with Web Workers from native apps, bypassing the JavaScript
thread to boost app performance and responsiveness, or to run tasks in the
background for offline support, push notifications, background sync, content
updates, and more. As a part of the work, the WebView2 team is adding support
for [Dedicated Workers](https://developer.mozilla.org/docs/Web/API/Web_Workers_API), [Shared Workers](https://developer.mozilla.org/docs/Web/API/SharedWorker), and [Service Workers](https://developer.mozilla.org/docs/Web/API/Service_Worker_API).

> While there are various types of web workers, our initial focus is on these three,

Reviewer: Please fix.

with plans to broaden our support for other workers in the future.

**What is Web Worker?** [Web Workers](https://developer.mozilla.org/docs/Web/API/Web_Workers_API) allow web
content to execute scripts in background threads, enabling tasks to run
independently of the user interface and facilitating communication between
threads using message passing. Web Workers can also make network requests using
fetch() or XMLHttpRequest APIs. Few examples of workers are

- **Dedicated Worker**: Dedicated Workers are a type of Web Worker that operates
  independently in the background of a web page, enabling parallel execution of
  tasks without blocking the main UI thread. They are created by specific web
  pages and communicate with them via message passing, making them ideal for
  offloading CPU-intensive operations and improving overall performance and
  responsiveness of web applications.

- **Shared Worker**: Shared Workers are a type of Web Worker that can be
  accessed by multiple scripts running in different windows, IFrames, or other
  contexts, as long as they are within the same domain as the worker. Unlike
  Dedicated Workers, which are tied to a specific web page, Shared Workers

>   require communication via an active port, making them slightly more complex to

Reviewer: Please fix.

- **Service Worker**: Service Workers serve as intermediary agents between web
  applications, the browser, and the network, functioning as proxy servers.
  Primarily, they facilitate the development of seamless offline experiences by
  caching resources and enabling web apps to continue functioning without an
  internet connection. Additionally, Service Workers intercept network requests,
  enabling customized responses based on network availability, and can update
  server-side assets. Moreover, they grant access to push notifications and
  background sync APIs, enhancing the functionality and interactivity of web
  applications.

# Description

We propose the introduction of APIs that enable the host application to oversee
workers. These APIs lay the groundwork for the development of additional APIs,
facilitating direct interaction between the host application and the workers,
and support background operations for service workers.

**DedicatedWorkerCreated**: This event, associated with CoreWebView2, is raised
when a web page initiates a dedicated worker. It grants access to the
CoreWebView2DedicatedWorker object, which contains information such as the

> script uri and name. Additionally, it provides a 'destroy' event that is raised

Reviewer: Please fix.

> just before this object is due for destruction.

Reviewer: Rename to `Destroying` its raised before the script worker or corewebview2 objects are destroyed.

the profile. This method gives a collection of
CoreWebView2ServiceWorkerRegistration objects, each encapsulating information
about a registered service worker, such as its script uri and scope uri.

**GetServiceWorkerRegistration**: The GetServiceWorkerRegistration method, part
of the CoreWebView2ServiceWorkerManager class, is used to retrieve a specific
service worker registration within the context of a WebView2 application for the
profile. This asynchronous method takes a scope uri as an argument and returns a
CoreWebView2ServiceWorkerRegistration object that encapsulates information about
the registered service worker with the given scope uri, such as its script uri. This
method is useful for accessing details of a specific service worker based on its
scope uri.

**SharedWorkerCreated**: This event, originating from the
CoreWebView2SharedWorkerManager associated with CoreWebView2Profile, is raised
when a shared worker is successfully created within the context of a WebView2
application for the profile. It grants access to the CoreWebView2SharedWorker
object, which contains information such as the script uri and name.

> Additionally, it provides a 'destroy' event that is raised just before this

Reviewer: Please fix.

object is due for destruction.

**GetSharedWorkers**: The GetSharedWorkers method, part of the
CoreWebViewSharedWorkerManager class, is used to retrieve all shared workers
created within the context of a WebView2 application for the profile. This
method gives a collection of CoreWebView2SharedWorkers objects.

# Background
This API specification provides a basic overview, including fundamental concepts
and introduces management APIs for interaction.

> Please note that while this specification does not add new APIs, future API

Reviewer 2: ```suggestion
Please note that future API
```

Reviewer: Please fix

specifications will introduce additional features for workers.

Stay tuned for updates to this spec and look out for new API specs that will
expand on the capabilities of workers. 

See work in progress Workers API specs

>  - [API Spec for SW Background Sync](https://github.com/MicrosoftEdge/WebView2Feedback/blob/api-periodic-sync-and-background-sync-draft/specs/PeriodicBackgoundSyncAndBackgroundSync.md)

Reviewer: Please fix.

# Examples
## Dedicated Worker
## Monitoring the Creation/Destruction of Dedicated Workers.
The following example demonstrates how to subscribe to the event that is raised
when a web page creates a dedicated worker. This event provides a
`CoreWebView2DedicatedWorker` object, enabling the host application to interact
with it.

This example also showcases the utilization of the upcoming PostMessage and

> WorkerMessageReceived APIs. These APIs enable efficient communication between

Reviewer: Discussed offline - this is part of a series of spec reviews and just this one on its own doesn't have practical sample code for real world.

the main thread and a dedicated worker.

## .NET/WinRT
```c#
    void PostMessageToWorker(CoreWebView2DedicatedWorker worker)
    {
        // Inside the worker, an event listener is set up for the 'message' event.
        // When a message is received from the host application, the worker
        // performs a fetch request. The host application also sets up an event
        // listener to receive messages from the worker. These messages could
        // contain fetched data or any error information from the worker's fetch
        // operation.
        worker.WorkerMessageReceived += (sender, args) =>
        {

>             var message = args.TryGetWorkerMessageAsString;

Reviewer 2: Is this really a property?

Reviewer: Please fix.

>             var message = args.TryGetWorkerMessageAsString;

Reviewer: Please fix.

            // This section of the code is dedicated to handling updates received
            // from the worker. Depending on the nature of these updates, you
            // might choose to perform various actions. For instance, you could
            // modify the user interface to reflect the new data, or you could
            // log the received data for debugging or record-keeping purposes.
        };

        // You can dispatch a message to the worker with the URL you want to

>         // fetch data from the host app.

Reviewer: We'll consider this in the future spec review for post message and workers.

>         worker.PostMessage("{ MyMessage: { type: 'FETCH_URL',url: 'https://example.com/data.json' }");

Reviewer: We'll consider this in the future spec review for post message and workers.

    }

>     void DedicatedWorkerCreatedExecuted(object target, ExecutedRoutedEventArgs e)

Reviewer 2: This sample looks like it's from a WPF test app. Copying from working code is great, but not knowing that context it's a confusing distraction. (And knowing the context would be info overload.) This could probably be updated to remove the test-isms

Reviewer: See above.

>     void DedicatedWorkerCreatedExecuted(object target, ExecutedRoutedEventArgs e)

Reviewer: Please fix: move the content of this method into its own method and have the menu item or whatever this method is for call into that one so we don't have to have this sample code UI leak into our sample app.

    {
        webView.CoreWebView2.DedicatedWorkerCreated += (sender, args) =>
        {
            CoreWebView2DedicatedWorker dedicatedWorker = args.Worker;

>             if(dedicatedWorker != null)

Reviewer 2: How could this be null?

Reviewer: See above.

>             if(dedicatedWorker != null)

Reviewer: Please fix: remove null check

            {

>                 dedicatedWorker.WorkerDestroyed += (sender, args) =>

Reviewer 2: ```suggestion
                dedicatedWorker.WorkerDestroyed += (sender2, args2) =>
```

Reviewer: Please fix.

                {
                  /*Cleanup on worker destruction*/
                };

                string scriptUri = dedicatedWorker.ScriptUri;

>                 LogMessage("Dedicated Worker has been created with the script located at the " + scriptUri);

Reviewer: Please update ref docs to say how URI normalization is applied or not with examples including hostnames. Please also check if we have existing docs on URI normalization that we can consolidate into one page of documentation.

>                 LogMessage("Dedicated Worker has been created with the script located at the " + scriptUri);

Author: Updated!

                // Send a message to the dedicated worker to offload fetch request from the host app.

>                 dedicatedWorker.PostMessageToWorker(dedicatedWorker);

Reviewer 2: Probably meant to send a message?

```suggestion
                dedicatedWorker.PostMessageToWorker(message);
```

Reviewer: See above.

>                 dedicatedWorker.PostMessageToWorker(dedicatedWorker);

Reviewer: Please fix.

            }
        };
    }
```
## Win32 C++

> ```cpp

Reviewer: Please fix

    void PostMessageToWorker(wil::com_ptr<ICoreWebView2DedicatedWorker> worker)
    {
        // Inside the worker, an event listener is set up for the 'message' event.
        // When a message is received from the host application, the worker performs
        // a fetch request. The host application also sets up an event listener
        // to receive messages from the worker. These messages could contain
        // fetched data or any error information from the worker's fetch operation.
        CHECK_FAILURE(worker->add_WorkerMessageReceived(
                    Callback<ICoreWebView2DedicatedWorkerMessageReceivedEventHandler>(

>                         [this](ICoreWebView2DedicatedWorker* sender,

Reviewer: Please fix: this doesn't appear to be necessary. Can have empty capture?

                            // This section of the code is dedicated to handling
                            // updates received from the worker. Depending on the
                            // nature of these updates, you might choose to perform
                            // various actions. For instance, you could modify
                            // the user interface to reflect the new data, or you
                            // could log the received data for debugging or
                            // record-keeping purposes.

                            return S_OK;
                        })
                        .Get(),
                    nullptr));

        // You can dispatch a message to the worker with the URL you want to fetch data from
        // the host app.
        CHECK_FAILURE(worker->PostMessage("{ MyMessage: {type: 'FETCH_URL',url: 'https://example.com/data.json' }"));
    }

    ScenarioDedicatedWorker::ScenarioDedicatedWorker(AppWindow* appWindow)
    : m_appWindow(appWindow), m_webView(appWindow->GetWebView())
    {
      m_webView2_25 = m_webView.try_query<ICoreWebView2_25>();

>       CHECK_FEATURE_RETURN_EMPTY(m_webView2_25);

Reviewer: Note for me. I will look into this in the future.

      CHECK_FAILURE(m_webView2_25->add_DedicatedWorkerCreated(
          Microsoft::WRL::Callback<ICoreWebView2DedicatedWorkerCreatedEventHandler>(
              [this](
                  ICoreWebView2* sender,

>                   ICoreWebView2DedicatedWorkerCreatedEventArgs* args)

Reviewer: Please fix.

>                   ICoreWebView2DedicatedWorkerCreatedEventArgs* args)

Reviewer 3: I think we have a lot of instances of this in our sample app that we may need to fix. @david-risney should we file an internal bug?

                      // Send a message to the dedicated worker to offload fetch
                      // request from the host app.
                      PostMessageToWorker(dedicatedWorker);
                  }

                  return S_OK;
              })
              .Get(),
          &m_dedicatedWorkerCreatedToken));
    }
```

## Service Worker
## Monitoring the Registration/Destruction of Service Workers

The following example demonstrates how to subscribe to the event that is raised
when a web page registers a service worker. This event provides a
`CoreWebView2ServiceWorkerRegistration` object, enabling the host application to
interact with the service workers via upcoming PostMessage and
WorkerMessageReceived APIs.

## .NET/WinRT
```c#
    void PostMessageToWorker(CoreWebView2ServiceWorker worker)
    {
        // The service worker communicates updates back to the host application,
        // such as when resource caching is complete. The host application can
        // listen for the WorkerMessageReceived event to receive these updates
        // from the service worker.

>         worker.WorkerMessageReceived += (sender, args) =>

Reviewer: See below in IDL section.

        {

>             var message = args.TryGetWorkerMessageAsString;

Reviewer: Future spec

            // Process the messages received from the worker. Depending on the
            // content of these messages, you might choose to log the data for
            // debugging, update the user interface to reflect changes, or raise
            // other actions within your application.
        };

>         // You can send messages to the service worker using its PostMessage method.

Reviewer: Please fix.

>         // The host application is sending a message to the worker to cache certain

Reviewer: Please fix

        // resources. The message is a JSON string that specifies the type of the
        // message and the payload of the message.

>         worker.PostMessage("{ MyMessage: { \"type\": \"CACHE_URLS\", \"payload\": [\"/styles/main.css\", \"/scripts/main.js\", \"/images/logo.png\"]} }");

Reviewer: Please fix

    }

    CoreWebView2ServiceWorkerManager ServiceWorkerManager_;
    void ServiceWorkerRegisteredExecuted(object target, ExecutedRoutedEventArgs e)
    {
        if (ServiceWorkerManager_ != nullptr)
        {
            ServiceWorkerManager_ = WebViewProfile.ServiceWorkerManager;
        }

>         ServiceWorkerManager_.ServiceWorkerRegistered += async (sender, args) =>

Reviewer: Please fix if unnecessary

        {

>             CoreWebView2ServiceWorkerRegistration serviceWorkerRegistration = args.ServiceWorkerRegistration;

Reviewer: Yes, matches web standard names

>             if(serviceWorkerRegistration != null)

Reviewer: Pleaes fix: remove

            {

>                 serviceWorkerRegistration.WorkerDestroyed += (sender, args) =>

Reviewer: Please fix: sample code doesn't match IDL (says Destroyed not WorkerDestroyed).

Rename to `Unregistering`

>                 {

Reviewer: See above

                    /*Cleanup on worker destruction*/
                };

                LogMessage("Service worker has been registered with a scope: " + serviceWorkerRegistration.ScopeUri);

                serviceWorkerRegistration.ServiceWorkerActivated += async (sender, args) =>
                {
                    CoreWebView2ServiceWorker serviceWorker = args.ActiveServiceWorker;

>                     serviceWorker.WorkerDestroyed += (sender, args) =>

Reviewer 2: Two events for worker destroyed, and listening to them both?

Reviewer: See above.

                          [this](
                              ICoreWebView2ServiceWorkerRegistration* sender,
                              IUnknown* args) -> HRESULT
                          {
                              /*Cleanup on worker registration destruction*/
                              return S_OK;
                          })
                          .Get(),
                    nullptr));

                    wil::unique_cotaskmem_string scopeUri;
                    CHECK_FAILURE(serviceWorkerRegistration->get_ScopeUri(&scopeUri));

                    CHECK_FAILURE(serviceWorkerRegistration->add_ServiceWorkerActivated(
                    Callback<ICoreWebView2ServiceWorkerActivatedEventHandler>(
                        [this](
                            ICoreWebView2ServiceWorkerRegistration* sender,
                            ICoreWebView2ServiceWorkerActivatedEventArgs* args) -> HRESULT
                        {
                            wil::com_ptr<ICoreWebView2ServiceWorker> serviceWorker;
                            CHECK_FAILURE(args->get_ActiveServiceWorker(&serviceWorker));
                        

>                             std::wstringstream message;

Reviewer: Please fix.

                  }

                  return S_OK;
              })
              .Get(),
          &m_serviceWorkerRegisteredToken));
    }
```

## Retrieving Service Worker Registrations.
The following example below illustrates how to retrieve and present details
about all service worker registrations linked to the WebView2 profile. This
yields a list of `CoreWebView2ServiceWorkerRegistration` objects, enabling the
host application to communicate directly with the service workers via post
messaging.

## .NET/WinRT
```c#
    private async void GetServiceWorkerRegistrationsExecuted(object target, ExecutedRoutedEventArgs e)
    {
        try
        {
            if (ServiceWorkerManager_ != null)
            {
                ServiceWorkerManager_ = WebViewProfile.ServiceWorkerManager;
            }
            IReadOnlyList<CoreWebView2ServiceWorkerRegistration> registrationList = await 
            ServiceWorkerManager_.GetServiceWorkerRegistrationsAsync();

            int registrationCount = registrationList.Count;
            StringBuilder messageBuilder = new StringBuilder();
            messageBuilder.AppendLine($"No of service workers registered: {registrationCount}");

>             for (int i = 0; i < registrationList.Count(); ++i)

Reviewer: Please fix

            {
                var scopeUri = registrationList[i].ScopeUri;

                messageBuilder.AppendLine($"Scope: {scopeUri}");

>             };

Reviewer: Please make sure ref docs say this: The event waits for next activation - caller must first check if already active worker is running.

            LogMessage(messageBuilder.ToString());
        }

>         catch (NotImplementedException exception)

Reviewer 2: Why is this necessary? Apps should never need to have catch blocks in their code. (Exceptions should only happen on app bugs.)

Reviewer: See above.

>         catch (NotImplementedException exception)

Reviewer: Please move try/catch out of sample code.

        {
            LogMessage("GetServiceWorkerRegistrationsAsync failed: " + exception.Message);
        }
    }
```
## Win32 C++
```cpp
    void ScenarioServiceWorkerManager::GetAllServiceWorkerRegistrations()
    {
        if (!m_serviceWorkerManager)
        {
            CreateServiceWorkerManager();
        }
        CHECK_FAILURE(m_serviceWorkerManager->GetServiceWorkerRegistrations(
            Callback<ICoreWebView2GetServiceWorkerRegistrationsCompletedHandler>(
                [this](
                    HRESULT error, ICoreWebView2ServiceWorkerRegistrationCollectionView*

>                                       workerRegistrationCollection) -> HRESULT

Reviewer: Please fix.

>                 {

Reviewer: Please fix.

                              i, &serviceWorkerRegistration));

                          wil::unique_cotaskmem_string scopeUri;
                          CHECK_FAILURE(serviceWorkerRegistration->get_ScopeUri(&scopeUri));

                          message << L"Scope: " << scopeUri.get();
                      }

                      LogMessage(std::wstring(message.str()));
                  }

                    return S_OK;
                })
                .Get()));
    }
```

## Retrieving Service Worker Registration for a Specific Scope.

The following example illustrates how to retrieve the
`CoreWebView2ServiceWorkerRegistration` associated with a specific scope. This
enables the host application to establish a communication channel with the
service worker, facilitating the exchange of messages.

## .NET/WinRT
```c#
    async void GetServiceWorkerRegisteredForScopeExecuted(object target, ExecutedRoutedEventArgs e)
    {
        var dialog = new TextInputDialog(
            title: "Scope of the Service Worker Registration",
            description: "Specify a scope uri to get the service worker",
            defaultInput: "");

>         if (dialog.ShowDialog() == true)

Reviewer: Please fix.  Avoid sync nested pump on UI thread

        {
            try
            {
                if (ServiceWorkerManager_ != null)
                {
                    ServiceWorkerManager_ = WebViewProfile.ServiceWorkerManager;
                }
                CoreWebView2ServiceWorkerRegistration registration = await

>                 ServiceWorkerManager_.GetServiceWorkerRegistrationAsync(dialog.Input.Text);

Reviewer: Please fix GetServiceWorkerRegistration docs to note about URI normalization.

>                 ServiceWorkerManager_.GetServiceWorkerRegistrationAsync(dialog.Input.Text);

Reviewer: Discussed differences between types of workers.

>                 ServiceWorkerManager_.GetServiceWorkerRegistrationAsync(dialog.Input.Text);

Author: Updated docs!

                if(registration != null)
                {

>                     CoreWebView2ServiceWorker worker = registration.ActiveServiceWorker;

Reviewer: Please ensure that it does: ActiveServiceWorker won't destroy until messages pumped and so on.

>                     CoreWebView2ServiceWorker worker = registration.ActiveServiceWorker;

Reviewer: If the worker instance exits the registration continues to exist.

-- NOTE this is where we stopped in the review meeting.

                                + std::wstring(ScriptUri.get()));
                            }
                            else
                            {
                                LogMessage(L"No service worker available");
                            }
                        else
                        {
                            LogMessage(L"No service worker registered for the scope: " + scopeUri);
                        }

                        return S_OK;
                    })
                    .Get()));
        }
    }
```

## Shared Worker
## Monitoring Shared Worker Creation and Destruction

The following example illustrates how the host application initiates a WebSocket
connection when a web page creates a shared worker, utilizing the Worker
upcoming PostMessage and WorkerMessageReceived APIs.

## .NET/WinRT
```c#
    void PostMessageToWorker(CoreWebView2SharedWorker worker)
    {
        // The shared worker communicates back to the host application with updates
        // from the WebSocket. The host application can monitor this event to receive
        // these updates from the shared worker.

>         worker.WorkerMessageReceived += (sender, args) =>

Author: Left and reviewed for future API's

        {
            var message = args.TryGetWorkerMessageAsString;

            // Record the updates from the WebSocket or modify the user interface
            // according to the requirements of the host application.
        };

        // Initiate the WebSocket connection by sending a message to the shared worker.

>         worker.PostMessage("{ MyMessage: {\"type\": \"INITIATE_WEBSOCKET_CONNECTION\", \"payload\": \"wss://example.com\"} }");

Author: Will add it in future  PostMessage APIs

    }

    CoreWebView2SharedWorkerManager SharedWorkerManager_;
    void SharedWorkerManagerExecuted(object target, ExecutedRoutedEventArgs e)
    {
        SharedWorkerManager_ = WebViewProfile.SharedWorkerManager;

        if(SharedWorkerManager_ != null)
        {
            SharedWorkerManager_.SharedWorkerCreated += (sender, args) =>
            {
                CoreWebView2SharedWorker sharedWorker = args.Worker;
                if(sharedWorker != null)
                {

>                   sharedWorker.WorkerDestroyed += (sender, args) =>

Author: Updated docs accordingly

                            message << L"ScriptUri: " << ScriptUri.get();
                            message << std::endl;
                        }
                    }

                    LogMessage(L"Shared workers" + std::wstring(message.str()));
                }
              return S_OK;
            })
            .Get()));
  }
```

# API Details
```
[uuid(29b994e5-0ac8-5430-89fa-5b4bb2091d8d), object, pointer_default(unique)]
interface ICoreWebView2_25 : IUnknown {
  /// Subscribe to this event that gets raised when a new dedicated worker is created.
  ///
  /// A Dedicated Worker is a type of web worker that allow you to run Javascript
  /// code in the background without blocking the main thread, making them useful
  /// for tasks like heavy computations, data processing, and parallel execution.
  /// It is "dedicated" because it is linked to a single parent document and cannot
  /// be shared with other scripts.
  ///
  /// This event is raised when a web application creates a dedicated worker using the
  /// `new Worker("/worker.js")` method. See the
  /// [Worker](https://developer.mozilla.org/docs/Web/API/Worker/Worker)
  /// for more information.

>   HRESULT add_DedicatedWorkerCreated(

Author: Haven't make any changes on this. @david-risney should we do this in the current spec?

  /// worker is executing. See the [MDN documentation]
  /// (https://developer.mozilla.org/docs/Web/API/Worker/scriptURL) for more information.
  ///
  /// The caller must free the returned string with `CoTaskMemFree`.  See
  /// [API Conventions](/microsoft-edge/webview2/concepts/win32-api-conventions#strings).
  [propget] HRESULT ScriptUri([out, retval] LPWSTR* value);

  /// Add an event handler for the `Destroyed` event that is raised when the
  /// worker object is destroyed.
  ///
  /// A worker object is destroyed when the worker script is terminated or when
  /// the `CoreWebView2DedicatedWorker` object is destroyed.
  HRESULT add_Destroyed(
      [in] ICoreWebView2DedicatedWorkerDestroyedEventHandler* eventHandler,
      [out] EventRegistrationToken* token);

  /// Removes an event handler previously added with `add_Destroyed`.
  HRESULT remove_Destroyed(
      [in] EventRegistrationToken token);
}

[uuid(ad6921d4-c416-5945-8437-2c97aeb76e6e), object, pointer_default(unique)]
interface ICoreWebView2Profile2 : IUnknown {
  /// Get the service worker manager to monitor service worker registrations and

>   /// interact with the service worker associated with the current profile.

Reviewer 4: ```suggestion
  /// interact with the service workers associated with the current profile.
```
Plural I assume. You can have multiple service workers associated with a profile.

  ///

>   /// The changes would apply to the context of the user profile. That is,

Author: Removed not necessary comments from the docs

  /// other WebViews under the same user profile could be affected.
  ///
  /// \snippet ScenarioServiceWorkerManager.cpp ServiceWorkerManager
  [propget] HRESULT ServiceWorkerManager([out, retval] ICoreWebView2ServiceWorkerManager** value);

  /// Get the shared worker manager to monitor shared worker creations and interact
  /// with the shared worker associated with the current profile.
  ///
  /// The changes would apply to the context of the user profile. That is, other
  /// WebViews under the same user profile could be affected.
  ///
  /// \snippet ScenarioSharedWorkerManager.cpp SharedWorkerManager
  [propget] HRESULT SharedWorkerManager([out, retval] ICoreWebView2SharedWorkerManager** value);
}

[uuid(4e07e562-8db7-5815-907b-6d89a253a974), object, pointer_default(unique)]
interface ICoreWebView2ServiceWorkerManager : IUnknown {

>   /// Adds an event handler for the `ServiceWorkerRegistered` event.

Reviewer 2: Adds some other things too. (I'd just remove this line)

  ///
  /// A ServiceWorker is a specific type of worker that takes a JavaScript file
  /// that can control the web-page/site that it is associated with,
  /// intercepting and modifying navigation and resource requests, and caching
  /// resources in a very granular fashion to give you complete control
  /// over how app behaves in certain situations.
  ///
  /// Service workers essentially act as proxy servers that sit between web applications,
  /// the browser, and the network (when available).
  /// Unlike Shared Workers, which have their own separate global scope, Service

>   ///Workers have no DOM access and run in a different context.

Reviewer 4: Confusing "Unlike..." sentence since the two parts do not contrast. Usually, "Unlike A, which does X, B does Y." you expect X and Y to be contrasting, but here the X and Y seem to talk about unrelated things.

  HRESULT remove_ServiceWorkerRegistered(
      [in] EventRegistrationToken token);

  /// Gets a list of the service worker registrations under the same profile.
  ///
  /// This method returns a list of `CoreWebView2ServiceWorkerRegistration` objects,
  /// each representing a service worker registration.
  ///
  /// This method corresponds to the `getRegistrations` method of the `ServiceWorkerContainer`
  /// object in the DOM which returns a Promise that resolves to an array of
  /// `ServiceWorkerRegistration` objects. See the [MDN documentation]
  /// (https://developer.mozilla.org/docs/Web/API/ServiceWorkerContainer/getRegistrations) for more information.
  HRESULT GetServiceWorkerRegistrations(
      [in] ICoreWebView2GetServiceWorkerRegistrationsCompletedHandler* handler
  );

  /// Gets the service worker registration object associated with the specified scope.
  /// If a service worker has been registered for the given scope, it gets the
  /// corresponding `CoreWebView2ServiceWorkerRegistration` object.
  ///
  /// The `scopeUri` parameter is a fully qualified URI that defines the range of URIs

>   /// a service worker can control. For example, if you registered a service worker
>   /// with a `scope` of '/app/' for an application at https://example.com/, you should
>   /// specify the full qualified URI i.e., https://example.com/app/ when calling this

Reviewer 4: ```suggestion
  /// a service worker can control. For example, instead of registering a service worker
  /// with a `scope` of '/app/' for an application at https://example.com/, you should
  /// specify the full qualified URI i.e., https://example.com/app/ when calling this
```
Clarify text. Original text was confusing, since it says "If you specified a scope of X, then you should specify a scope of Y", which is closing the barn door after the horse has bolted.

Reviewer 4: I assume we return a failure if the scope is not fully qualified?

Author: Updated docs!

      [in] IUnknown* args);
}

[uuid(0c0b03bd-ced2-5518-851a-3d3f71fb5c4b), object, pointer_default(unique)]
interface ICoreWebView2ServiceWorkerRegistration : IUnknown {
  /// The `scopeUri` parameter is a fully qualified URL that defines the range of URLs
  /// a service worker can control.
  ///
  /// By default, the `scopeUri` is the directory where the service worker script resides.
  /// For example, if the script is at https://example.com/app/sw.js, the default scopeUri is
  /// https://example.com/app/, and the service worker can control all pages under this URI.
  ///
  /// During the registration of the service worker, the scope is set relative to the
  /// application's base URI. For example, if you register a service worker with a scope
  /// of /app/ for an application at https://example.com/, the scope is relative to the
  /// base URI during registration.
  ///
  /// After the service worker is registered, the `scope` is stored and returned as a
  /// fully qualified URI, such as https://example.com/app/. This URI includes the origin
  /// (scheme, host, and optionally port) and the path. If the origin or scope includes

>   /// an internationalized domain name (IDN), it should be represented in punycode format.

Author: Updated

  /// In such a case, the existing CoreWebView2ServiceWorker object is destroyed, and this event is
  /// raised with a new CoreWebView2ServiceWorker object representing the updated service worker.
  /// The active service worker is the one that receives fetch and message events for the pages it controls.
  /// See the [Service Worker](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerRegistration/active)
  /// documentation for more information.
  /// 
  HRESULT add_ServiceWorkerActivated(
      [in] ICoreWebView2ServiceWorkerActivatedEventHandler* eventHandler,
      [out] EventRegistrationToken* token);

  /// Removes an event handler previously added with `add_ServiceWorkerActivated`.
  HRESULT remove_ServiceWorkerActivated(
      [in] EventRegistrationToken token);

  /// Add an event handler for the `Destroyed` event that is raised when the worker registration is
  /// unregistered using the JS api `registration.unregister()` method or when the
  /// `CoreWebView2ServiceWorkerRegistration` object is destroyed. See the
  /// [Unregister](https://developer.mozilla.org/docs/Web/API/ServiceWorkerRegistration/unregister)
  /// for more information.

>   HRESULT add_Destroyed(

Reviewer 2: This is both on the SW registration and the SW itself for convenience?

        Windows.Foundation.IAsyncOperation<IVectorView<CoreWebView2ServiceWorkerRegistration>> GetServiceWorkerRegistrationsAsync();

        Windows.Foundation.IAsyncOperation<CoreWebView2ServiceWorkerRegistration> GetServiceWorkerRegistrationAsync(String ScopeUri);
    }

    runtimeclass CoreWebView2ServiceWorkerRegisteredEventArgs
    {
        CoreWebView2ServiceWorkerRegistration ServiceWorkerRegistration { get; };
    }

    runtimeclass CoreWebView2ServiceWorkerRegistration
    {
        String ScopeUri { get; };

        event Windows.Foundation.TypedEventHandler<CoreWebView2ServiceWorkerRegistration, IInspectable> Destroyed;

        CoreWebView2ServiceWorker ActiveServiceWorker { get; };

        event Windows.Foundation.TypedEventHandler<CoreWebView2ServiceWorkerRegistration, CoreWebView2ServiceWorkerActivatedEventArgs> ServiceWorkerActivated;
    }

    runtime CoreWebView2ServiceWorkerActivatedEventArgs
    {
        CoreWebView2ServiceWorker ActiveServiceWorker { get; };
    }

    runtimeclass CoreWebView2ServiceWorker
    {
        String ScriptUri { get; };

>         CoreWebView2ServiceWorkerState State { get; };

Author: Yes, will remove. Thanks!

>         CoreWebView2ServiceWorkerState State { get; };

Reviewer 2: So there's no `State` property? Is the notion of an "active" worker gone too?
